src/metabase/query_processor/store.clj

```diff
@@ -310,39 +84,217 @@
+                          {:metadata-provider new-provider}
+                          e))))
+      (store-miscellaneous-value! [::metadata-provider] new-provider))))
+
+(defn do-with-metadata-provider
+  "Implementation for [[with-metadata-provider]]."
+  [database-id-or-metadata-provider thunk]
+  (if-not (initialized?)
+    (binding [*store* (atom {})]
+      (do-with-metadata-provider database-id-or-metadata-provider thunk))
+    (do
+      (maybe-set-metadata-provider! database-id-or-metadata-provider)
+      (thunk))))
 
-    (table [_this table-id]
-      (some-> (table table-id) (update-keys u/->kebab-case-en) (assoc :lib/type :metadata/table)))
+(defmacro with-metadata-provider
+  "Execute `body` with an initialized QP store and metadata provider bound. You can either pass
+  a [[metabase.lib.metadata.protocols/MetadataProvider]] directly, or pass a Database ID, for which we will create
+  a [[metabase.lib.metadata.jvm/application-database-metadata-provider]].
 
-    (field [_this field-id]
-      (some-> (field field-id) (update-keys u/->kebab-case-en) (assoc :lib/type :metadata/column)))
+  If a MetadataProvider is already bound, this is a no-op."
+  {:style/indent [:defn]}
+  [database-id-or-metadata-provider & body]
+  `(do-with-metadata-provider ~database-id-or-metadata-provider (^:once fn* [] ~@body)))
 
-    (card [_this _card-id] nil)
-    (metric [_this _metric-id] nil)
-    (segment [_this _segment-id] nil)
-    (tables [_metadata-provider] nil)
-    (fields [_metadata-provider _table-id] nil)
-    (metrics [_metadata-provider _table-id] nil)
+(def ^:private IDs
+  [:maybe
+   [:or
+    [:set ::lib.schema.common/positive-int]
+    [:sequential ::lib.schema.common/positive-int]]])
+
+(defn- missing-bulk-metadata-error [metadata-type id]
+  (ex-info (tru "Failed to fetch {0} {1}" (pr-str metadata-type) (pr-str id))
+           {:status-code       400
+            :type              qp.error-type/invalid-query
+            :metadata-provider (metadata-provider)
+            :metadata-type     metadata-type
+            :id                id}))
+
+(mu/defn bulk-metadata :- [:maybe [:sequential [:map
+                                                [:lib/type :keyword]
+                                                [:id ::lib.schema.common/positive-int]]]]
+  "Fetch multiple objects in bulk. If our metadata provider is a bulk provider (e.g., the application database metadata
+  provider), does a single fetch with [[lib.metadata.protocols/bulk-metadata]] if not (i.e., if this is a mock
+  provider), fetches them with repeated calls to the appropriate single-object method,
+  e.g. [[lib.metadata.protocols/field]].
+
+  The order of the returned objects will match the order of `ids`, and the response is guaranteed to contain every
+  object referred to by `ids`. Throws an exception if any objects could not be fetched."
+  [metadata-type :- [:enum :metadata/card :metadata/column :metadata/metric :metadata/segment :metadata/table]
```

Comment:
I don't think it's defined anywhere else, and at any rate the types of metadata supported by the QP store are not necessarily the same as what might be supported elsewhere. So I'd rather leave this here for now since it's QP-Store-specific